doc/website/blog: atlas gorm provider view blogpost (#2804)
ariga/atlas: doc/website/blog/2024-05-24-atlas-gorm-provider-view.mdx

---
title: "Supercharge your GORM migration with Atlas: Views Support"
authors: luantranminh
tags: [gorm, views, migration]
---

:::info TL;DR

> Lack of full support for views in GORM migration? Atlas is here to help! With the new release of 
> Atlas GORM Provider, you can now not only define and manage migrations for tables but also views with ease.
> 
> [See an example](#demo-time)

Reviewer: ```suggestion
[GORM](https://gorm.io), a popular ORM for Go can easily query SQL views, but managing them has traditionally
been [an issue](https://github.com/go-gorm/gorm/issues/4966). With the [latest release](https://github.com/ariga/atlas-provider-gorm/releases/tag/v0.4.0) of 
[Atlas GORM Provider](https://atlasgo.io/guides/orms/gorm), you can now easily manage views for your GORM application. 

[See an example](#demo-time)
```

:::

## Introduction
 

> Views are a powerful database feature: they are virtual table representing the result of a query. 
> Many teams use them to simplify complex queries, encapsulate logic, and present
>  a consistent interface to users, abstracting the underlying data structures.
> 
> > Making liberal use of views is a key aspect of good SQL database design.
> >
> > — <cite>**Postgres documentation**</cite>

Reviewer: ```suggestion
> Making liberal use of views is a key aspect of good SQL database design.
>
> — <cite>**Postgres documentation**</cite>

Views are a powerful database feature: they are virtual table representing the result of a query. 
Many teams use them to simplify complex queries, encapsulate logic, and present
 a consistent interface to users, abstracting the underlying data structures.
```

## The current landscape

Despite the benefits, views are often underutilized in many projects. Defining and managing database views can be tedious
 and error-prone, especially when it comes to migration. 

> GORM is a fantastic ORM library that supports auto creation for database schema 

Reviewer: Show GORMs current API for managing views and discuss it.

```bash
atlas login
```

Install the provider by running:

```bash
go get -u ariga.io/atlas-provider-gorm
```

#### Step 1: Create a GORM Application

Models are defined using normal structs. For views, we define a struct and 
implement the [ViewDefiner]([https://pkg.go.dev/ariga.io/atlas-provider-gorm/gormschema#ViewDefiner]) interface. The `ViewDef(dialect string)` method receives the 
`dialect` argument to determine the SQL dialect to generate the view. It is helpful for generating the view definition for different SQL dialects if needed.

Let's create a file that will contain our database models. We will call it `models/models.go`

```go title="models/models.go"
package models

import (
  "ariga.io/atlas-provider-gorm/gormschema"
  "gorm.io/gorm"
)

// User is a regular gorm.Model stored in the "users" table.
type User struct {
  gorm.Model
  Name   string
  Age    int
  Gender string
}  

// WorkingAgedUser is mapped to the VIEW definition below.
type WorkingAgedUser struct {
  Name string
  Age  int
}

func (WorkingAgedUser) ViewDef(dialect string) []gormschema.ViewOption {
    return []gormschema.ViewOption{

>         gormschema.CreateStmt(`

Reviewer: This is the core part of the blogpost where you reveal the new tech. 

Technical blog readers are skeptical and curious. 

Show off the different alternatives - using GORM syntax (first. this is the familiar API to the user), then alternatively show providing the query. Mention that it is DB specific, and show how to use a switch case statement where necessary.

Author: > Mention that it is DB specific, and show how to use a switch case statement where necessary.

Hey @rotemtam, I need a little help with this part. Should I mention it as the main example or keep my original example and include the switch case as an additional section?

The reason is that I want to show the code differences later, so I use multiple lines for the SQL statement. Including the switch case in the example will make it extremely verbose

Reviewer: You can use a `<Tabs>` component to make examples more condensed, if that works

* **Standalone**: If your views and models are in the same package, you can use the provider directly to load your GORM schema into Atlas.
* **Go Program**: If you have them defined in different packages, you can use the provider as a library in your Go program to load your GORM schema into Atlas.

Since all of our models are in the same package, it's pretty handy to use 
the **Standalone** mode. But if you're curious, you can also try the **Go Program** mode with more detail in the [GORM Guide](https://atlasgo.io/guides/orms/gorm).

In your project directory, create a new file named `atlas.hcl` with the following contents:

```hcl title="atlas.hcl"
data "external_schema" "gorm" {
  program = [
    "go",
    "run",
    "-mod=mod",
    "ariga.io/atlas-provider-gorm",
    "load",
    "--path", "./models" // path to your models
    "--dialect", "mysql", // | postgres | sqlite | sqlserver
  ]
}

env "gorm" {
  src = data.external_schema.gorm.url
  dev = "docker://mysql/8/dev" // the dev-database needs to be mapped to the same dialect above
  migration {
    dir = "file://migrations"
  }
  format {
    migrate {
      diff = "{{ sql . \"  \" }}"
    }
  }
}
```

> :::info Running with dialects
> 
> To run Atlas with the `mysql`, `postgres`, or `sqlserver` dialects, make sure your Docker is up and running.
> 
> :::

Reviewer: ```suggestion
:::info Using docker://

If you use the `docker://` driver for spinning up your [Dev Database](https://atlasgo.io/concepts/dev-database) 
be sure that Docker is running locally on your machine first. 

:::
```

>
> — <cite>**Postgres documentation**</cite>

Views are a powerful database feature: they are virtual table representing the result of a query. 
Many teams use them to simplify complex queries, encapsulate logic, and present
 a consistent interface to users, abstracting the underlying data structures.

## The current landscape

Despite their numerous benefits, views are often underutilized in many applications. Softwares like ORMs play a key role 
in educating developers and increasing the adoption of views. However, many ORMs only provide partial support for views.

Let's explore how GORM, one of the most popular ORMs in Go, enables users to create and utilize views in 
their applications. 

First, we need to define the query that will back our view, and then use 
the GORM [Migrator](https://pkg.go.dev/gorm.io/gorm/migrator#Migrator) interface to create the view.

```go
query := db.Model(&User{}).Select("id, name, age").Where("age BETWEEN 18 AND 60")

db.Migrator().CreateView("working_aged_users", gorm.ViewOption{Query: query})
// CREATE VIEW working_aged_users AS SELECT id, name, age FROM users WHERE age BETWEEN 18 AND 60
```

> Next, to represent our view, we need to define a corresponding struct:

Author: @rotemtam  Does "_additional_" make the "struct" sound redundant? Since our approach also requires a struct as well

> Next, to represent our view, we need to define a corresponding struct:

Reviewer: I think it highlights the complexity. you can ignore you don't like it

---
title: "Manage and Query SQL Views with GORM and Atlas"
authors: luantranminh
tags: [gorm, views, migration]
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

:::info TL;DR

[GORM](https://gorm.io), a popular ORM for Go can easily query SQL views, but managing them has traditionally
been [an issue](https://github.com/go-gorm/gorm/issues/4966). With the [latest release](https://github.com/ariga/atlas-provider-gorm/releases/tag/v0.4.0) of 
[Atlas GORM Provider](https://atlasgo.io/guides/orms/gorm), you can now easily manage views for your GORM application. 

[See an example](#demo-time)

:::

## Introduction
 
> Making liberal use of views is a key aspect of good SQL database design.
>
> — <cite>[**Postgres documentation**](https://www.postgresql.org/docs/current/tutorial-views.html#TUTORIAL-VIEWS)</cite>

Views are a powerful database feature: they are virtual table representing the result of a query. 
Many teams use them to simplify complex queries, encapsulate logic, and present

>  a consistent interface to users, abstracting the underlying data structures.

Reviewer: ```suggestion
a consistent interface to users, abstracting the underlying data structures.
```

## Using Views with ORMs

Despite their numerous benefits, views are often underutilized in many applications. Specifically,
many ORMs provide partial support for views.

> This is also the case with GORM, one of the most popular ORMs in Go.Let's see how GORM users can integrate views into their applications today: 

Reviewer: ```suggestion
This is also the case with GORM, one of the most popular ORMs in Go. Let's see how GORM users can
integrate views into their applications today: 
```

Reviewer: spacing
